pkg/compact/downsample/downsample.go

```diff
@@ -256,41 +318,124 @@ type aggregator struct {
+	f.max = -math.MaxFloat64
 }
 
-func (a *aggregator) add(v float64) {
-	if a.total > 0 {
-		if v < a.last {
+func (f *floatAggregator) add(s sample) {
+	if f.total > 0 {
+		if s.v < f.last {
 			// Counter reset, correct the value.
-			a.counter += v
-			a.resets++
+			f.counter += s.v
+			f.resets++
 		} else {
 			// Add delta with last value to the counter.
-			a.counter += v - a.last
+			f.counter += s.v - f.last
 		}
 	} else {
 		// First sample sets the counter.
-		a.counter = v
+		f.counter = s.v
 	}
-	a.last = v
+	f.last = s.v
 
-	a.sum += v
-	a.count++
-	a.total++
+	f.sum += s.v
+	f.count++
+	f.total++
 
-	if v < a.min {
-		a.min = v
+	if s.v < f.min {
+		f.min = s.v
 	}
-	if v > a.max {
-		a.max = v
+	if s.v > f.max {
+		f.max = s.v
 	}
 }
 
+func (f *floatAggregator) processedSamples() int {
+	return f.total
+}
+
+type histogramAggregator struct {
+	total         int                       // Total histograms processed.
+	count         int                       // Histograms in current window.
+	sum           *histogram.FloatHistogram // Value sum of current window (for gauge histograms).
+	counter       *histogram.FloatHistogram // Total counter state since beginning (for counter histograms).
+	resets        int                       // Number of counter resets since beginning.
+	previous      *histogram.FloatHistogram // Previously added value.
+	schema        int32                     // Smallest schema in the batch that's being aggregated.
+	resetDetected bool                      // Whether a reset was detected since the beginning.
+}
+
+func newHistogramAggregator(schema int32) *histogramAggregator {
+	return &histogramAggregator{
+		schema: schema,
+	}
+}
+
+func (h *histogramAggregator) reset() {
+	h.count = 0
+	h.sum = nil
+}
+
+func (h *histogramAggregator) add(s sample) {
+	fh := s.fh
+	if fh.Schema < h.schema {
+		panic("schema must be greater or equal to aggregator schema")
+	}
+
+	// A schema increase is treated as a reset, so we need to preserve
+	// the original histogram in case the schema is adjusted.
+	oFh := fh
+	// If schema of the sample is greater than the
+	// aggregator schema, we need to reduce the resolution.
+	if fh.Schema > h.schema {
+		fh = fh.CopyToSchema(h.schema)
+	}
+
+	if h.total > 0 {
+		if fh.CounterResetHint != histogram.GaugeType && oFh.DetectReset(h.previous) {
+			// Counter reset, correct the value.
+			h.counter.Add(fh)
+			h.resets++
+			h.resetDetected = true
+		} else {
+			// Add delta with previous value to the counter.
+			h.counter.Add(fh.Copy().Sub(h.previous))
+		}
+	} else {
+		// First sample sets the counter.
+		h.counter = fh.Copy()
+	}
+
+	if h.sum == nil {
+		if h.resetDetected {
+			h.sum = h.counter.Copy()
+		} else {
+			h.sum = fh.Copy()
+		}
+		// This needs to be h gauge histogram, otherwise reset detection will be triggered
+		// when appending the aggregated chunk and histogram.count < appender.count.
+		h.sum.CounterResetHint = histogram.GaugeType
```

Comment:
Isn't the result that this is always set when adding multiple histograms? Maybe we can move this out of the if?